x/metadata/types/address_test.go

```diff
@@ -420,6 +451,148 @@ func (s *AddressTestSuite) TestVerifyMetadataAddressFormat() {
+			hrp:    PrefixScope,
+			expErr: "invalid scope metadata address MetadataAddress{0x0, 0x1, 0x2}: incorrect address length (expected: 17, actual: 3)",
+		},
+		{
+			name:   "invalid session id",
+			ma:     MetadataAddress{SessionKeyPrefix[0], 0x3, 0x4},
+			hrp:    PrefixSession,
+			expErr: "invalid session metadata address MetadataAddress{0x1, 0x3, 0x4}: incorrect address length (expected: 33, actual: 3)",
+		},
+		{
+			name:   "invalid record id",
+			ma:     MetadataAddress{RecordKeyPrefix[0], 0x5, 0x6},
+			hrp:    PrefixRecord,
+			expErr: "invalid record metadata address MetadataAddress{0x2, 0x5, 0x6}: incorrect address length (expected: 33, actual: 3)",
+		},
+		{
+			name:   "invalid scope spec id",
+			ma:     MetadataAddress{ScopeSpecificationKeyPrefix[0], 0x7, 0x8},
+			hrp:    PrefixScopeSpecification,
+			expErr: "invalid scope specification metadata address MetadataAddress{0x4, 0x7, 0x8}: incorrect address length (expected: 17, actual: 3)",
+		},
+		{
+			name:   "invalid contract spec id",
+			ma:     MetadataAddress{ContractSpecificationKeyPrefix[0], 0x9, 0xa},
+			hrp:    PrefixContractSpecification,
+			expErr: "invalid contract specification metadata address MetadataAddress{0x3, 0x9, 0xa}: incorrect address length (expected: 17, actual: 3)",
+		},
+		{
+			name:   "invalid record spec id",
+			ma:     MetadataAddress{RecordSpecificationKeyPrefix[0], 0xb, 0xc},
+			hrp:    PrefixRecordSpecification,
+			expErr: "invalid record specification metadata address MetadataAddress{0x5, 0xb, 0xc}: incorrect address length (expected: 33, actual: 3)",
+		},
+	}
+
+	validCases := []struct {
+		hrp string
+		ma  MetadataAddress
+	}{
+		{hrp: PrefixScope, ma: ScopeMetadataAddress(newUUID("1"))},
+		{hrp: PrefixSession, ma: SessionMetadataAddress(newUUID("2"), newUUID("3"))},
+		{hrp: PrefixRecord, ma: RecordMetadataAddress(newUUID("4"), newUUID("5").String())},
+		{hrp: PrefixScopeSpecification, ma: ScopeSpecMetadataAddress(newUUID("6"))},
+		{hrp: PrefixContractSpecification, ma: ContractSpecMetadataAddress(newUUID("7"))},
+		{hrp: PrefixRecordSpecification, ma: RecordSpecMetadataAddress(newUUID("8"), newUUID("9").String())},
+	}
+	for _, vc := range validCases {
+		for _, typeVC := range validCases {
+			tc := testCase{
+				name: fmt.Sprintf("valid %s: want %s", vc.hrp, typeVC.hrp),
+				ma:   vc.ma,
+				hrp:  typeVC.hrp,
+			}
+			if vc.hrp != typeVC.hrp {
+				tc.expErr = fmt.Sprintf("invalid %s id \"%s\": wrong type", getNameForHRP(tc.hrp), tc.ma.String())
+			}
+			tests = append(tests, tc)
+		}
+	}
+
+	for _, tc := range tests {
+		s.Run(tc.name, func() {
+			var err error
+			testFunc := func() {
+				err = VerifyMetadataAddressHasType(tc.ma, tc.hrp)
+			}
+			s.Require().NotPanics(testFunc, "VerifyMetadataAddressHasType(%q, %q)", tc.ma, tc.hrp)
+			assertions.AssertErrorValue(s.T(), err, tc.expErr, "error from VerifyMetadataAddressHasType(%q, %q)", tc.ma, tc.hrp)
+		})
+	}
+}
```

Comment:
The only benefit to hard coding those invalid id errors is to give a person something to look at. But it doesn't matter in this case. And by making them dynamic like that, I get all the different combinations out of 6 lines instead of 36.